lib/defopt.py

```diff
@@ -246,6 +246,38 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         sys.exit(e)
 
 
+def _recurse_functions(funcs, subparsers):
```

Comment:
Would it not be easier if you normalized to dicts at the very top, i.e. `if not isinstance(funcs, Mapping): funcs = {func.__name__.replace(...): func for func in funcs}`?  Especially because otherwise the `for func in funcs` iteration can have different meanings depending on the type of funcs.

Reply:
Ya, so basically convert any list to a dictionary and then assume we always have a dictionary in the rest of the function.  I can do this.  Gotta run for a bit, but I'll make this change soon!

Reply:
Ok, I made this change.  I capture the `AttributeError` that is raised if trying to access __name__ for a list so that I can raise the ValueError instead.  Alternatively, we could leave it as the AttributeError, although then it doesn't give the user any reason on what they did wrong.